src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSTempPECompiler.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.LanguageServices.ProjectSystem;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.CPS
+{
+    [Export(typeof(ITempPECompiler))]
+    internal class CPSTempPECompiler : ITempPECompiler
+    {
+        public async Task<bool> CompileAsync(IWorkspaceProjectContext context, string outputFileName, string[] filesToInclude, CancellationToken cancellationToken)
+        {
+            if (filesToInclude == null || filesToInclude.Length == 0)
+            {
+                throw new ArgumentException(nameof(filesToInclude), "Must specify some files to compile.");
+            }
+            if (outputFileName == null)
+            {
+                throw new ArgumentException(nameof(outputFileName), "Must specify an output file name.");
+            }
+
+            var project = ((CPSProject)context).GetProjectSnapshot();
+
+            // Allow for faster checking because projects could be very large
+            var files = new HashSet<string>(filesToInclude, StringComparer.OrdinalIgnoreCase);
+
+            // Remove all files except the ones we care about
+            var documents = project.Documents;
+            foreach (var document in documents)
+            {
+                if (!files.Contains(document.FilePath))
+                {
+                    project = project.RemoveDocument(document.Id);
```

Comment:
Just an FYI: a large number of removes in a loop might be fairly expensive if you're doing large solutions and only asking us to build a small number of files. If you see this as a hotspot in the trace, let's add a RemoveDocuments() call. We've had other places too where it would be useful.

Reply:
(and to be clear: I wouldn't deal with this until you have perf traces showing this to be a problem)

Reply:
I definitely expect this to be a problem for large solutions, and in fact I was looking for a RemoveDocuments() method when I first started writing this. The largest Windows Forms project I worked on at my previous job had almost 600 C# source files and only one of them would have been compiled into a TempPE dll.

Reply:
600?  That's nothing.  The perf bug I looked at last year with finding entry points had to do with a winforms project with over 6,000 source files...